app/models/mixins/supports_feature_mixin.rb

```diff
@@ -99,24 +87,46 @@ def unsupported
   class_methods do
     # This is the DSL used a class level to define what is supported
     def supports(feature, &block)
-      define_supports_feature_methods(feature, &block)
+      self.supports_features = supports_features.merge(feature.to_sym => block || true)
```

Comment:
My only concern here is that I think this needs to be made thread-safe.  This is because we could lazy load classes at run-time in different threads (think 2 different requests coming that each need to lazy load things).

I know that lazy loading has bit us in the past with this - see the way the `unsupported` method is defined to use a Concurrent::Hash for this reason.

Reply:
This is the Rails pattern for defining class level configuration that inherits from parent classes and then inherit columns from parents and allow overriding.

- define class variable https://github.com/rails/rails/blob/7-0-stable/activerecord/lib/active_record/attributes.rb#L11
- append to class variable https://github.com/rails/rails/blob/7-0-stable/activerecord/lib/active_record/attributes.rb#L231
- use variable to set methods https://github.com/rails/rails/blob/7-0-stable/activerecord/lib/active_record/attributes.rb#L265

We use this same pattern for virtual_attributes:

- https://github.com/kbrock/activerecord-virtual_attributes/blob/master/lib/active_record/virtual_attributes.rb#L41
- https://github.com/kbrock/activerecord-virtual_attributes/blob/master/lib/active_record/virtual_attributes.rb#L62


Since these values are set at class load time, and Rails has a class loader mutex to protect the other loaders, I am pretty sure we are in the clear.

Reply:
`supported_features`, similar to rails attribute API, only modify the class at load time and should be protected by the rails class loader mutex.

`unsupported` is constantly modified at runtime. While I am not happy with this race condition, this PR focuses on removing the extra methods that were causing the infinite backtraces. In the future we can hopefully drop `unsupported` and use `supported_feature` instead.

Reply:
I created #22898 as a place holder for the next phase

Reply:
This code does a get->modify->set, and another thread can get in the middle.  In the past, 2 threads could lazy-load 2 separate classes, particularly in the API.

> only modifies the class at load time and should be protected by the rails class loader mutex.

@jrafanie Can you comment here?  Not sure if this is now completely handled by zeitwerk or not, but this was a problem in the past and is why we had to wrap all of those class loading places in the special Rails handler, particularly in the API.

Reply:
hmm, there is [discussion of thread safety for code reloading](https://github.com/fxn/zeitwerk?tab=readme-ov-file#thread-safety).  I'm wondering if we need to implement something similar to [what we had previously](https://github.com/ManageIQ/manageiq/pull/22801) for dev mode reloading only.  I think autoloading in production mode is fine as long as it's after the application is initialized/rails loaded.

Reply:
> This code does a get->modify->set, and another thread can get in the middle. In the past, 2 threads could lazy-load 2 separate classes, particularly in the API.

Yes, The 'get, modify, set' of `unsupported[]` is bad. This is the existing behavior. 

The first commit's goal is to just consolidate the existing logic.
Rails uses this patter when defining methods: `def field_name ; attr_reader(:field_name) ; end` instead of putting the actual gook into the defined method. (not always, but generally) 

```diff
  def supports(feature, &block)
    define_method("supports_#{feature}?") do
+     common_method(feature)
+   end
+ end
+
+   def common_method(feature) # sorry rubocop for indentation
      unsupported.delete(feature)
      if block
        instance_eval(&block) # gook (not for the class level)
      else
        unsupported.add(feature) # some kind of modify
      end
      !!unsupported[feature]
    end
- end
```

I can see if there is a way to make this fact more obvious.
But again, I wasn't as concerned about fixing the code knowing that #22898 will remove the `clear ; modify ; get` to a just plain `get`.


>> only modifies the class at load time
>> [...] this was a problem in the past and is why we had to wrap all of those class loading places in the special Rails handler

There have been issues when one class referenced another class at load time. `virtual_delegates` used to have this problem but we changed it to lazy reference the other class. `has_many` also has this problem, and why we use `class_name=String` rather than class references.

In our case, we are following the example from the `attributes()` api. It defines Symbols/Strings that are put into a `Hash` at load time and never changed after that.

Reply:
Ok discussed with @kbrock and I was mistakenly thinking that the class_attribute would be _shared_, so that child classes would be pushing values and could clobber each other.  However, what's actually happening is the parent class loads first, then each child creates a copy of the parent class' feature and then inserts their own into the copy.

So this is fine :+1: